### PR TITLE
Design:  ordered creation and deletion of composed resources

### PR DESCRIPTION
**Pull request by stevendborrelli**
### Description of your changes

Design proposal for Crossplane to sequence the creation and deletion of a Composite Resource's composed resources, from a dependency graph the function pipeline declares.

## Background

An overview of existing methods of sequencing resources is at:  <https://github.com/stevendborrelli/crossplane/blob/e43e3fecfe10a3b0fa6fd6797422b1bd7b7db28a/design/assets/design-doc-composed-resource-ordering/background.md>

## Prototype and Testing

A Prototype PR #7842 was created to flesh out this proposal and find edge cases. It includes a number of test sequences.

A testing guide is available at <https://github.com/stevendborrelli/crossplane/blob/composed-resource-ordering-prototype/design/assets/design-doc-composed-resource-ordering/notes-testing-composed-resource-ordering.md>

This prototype depends on:

- Crossplane runtime <https://github.com/crossplane/crossplane-runtime/pull/1154>
- While not a dependency, the PR <https://github.com/crossplane-contrib/provider-nop/pull/27> add the ability to delay resource deletion for teardown testing.
- Function-ordering <https://github.com/stevendborrelli/function-ordering>, a fork of function-sequencer

Fixes #

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Read and followed Crossplane's [AI policy] and confirmed a human stands behind this PR.
- [x] Run `./nix.sh flake check` to ensure this PR is ready for review.
- ~[ ] Added or updated unit tests.~
- ~[ ] Added or updated e2e tests.~
- ~[ ] Linked a PR or a [docs tracking issue] to [document this change].~
- ~[ ] Added `backport release-x.y` labels to auto-backport this PR.~
- [x] Followed the [API promotion workflow] if this PR introduces, removes, or promotes an API.

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/main/contributing
[AI policy]: https://github.com/crossplane/crossplane/blob/main/AI_POLICY.md
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/main/contributing#checklist-cheat-sheet
[API promotion workflow]: https://github.com/crossplane/crossplane/blob/main/contributing/guide-api-promotion.md

### CODE REVIEW

**Comment by stevendborrelli** [comment]
Six of these were right and I've fixed them in 6f7a7cf.  I've resolved the relevant comments.

File: design/assets/design-doc-composed-resource-ordering/background.md [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+* **Deletion ordering costs a Kubernetes Object per protected Resource.**
+  Expressing teardown order through `Usage` objects means one `Usage` per
+  dependency pair: each with its own reconcile, finalizer, and admission round
+  trip. It also requires deleting the XR with `--cascade=foreground`, which is
+  the one propagation mode that defeats function-controlled deletion (see
+  "Interaction with Function-Controlled Deletion" below).
+
+In a complex project like
+[modelplane](https://github.com/modelplaneai/modelplane) a single deployment can
+contain over a dozen `Usage` objects on the cluster.
```
Comment (bobh66) [inline design/assets/design-doc-composed-resource-ordering/background.md]: While accurate I think this understates the impact.  We have deployments that contain 100+ `Usage` resources to maintain proper deletion ordering, and several composites that contain dozens of `Usages` just within the composition.  This leads to hundreds/thousands of `Usage` resources managed by a single Crossplane deployment.

File: design/design-doc-composed-resource-ordering.md [hunk 2]
```diff
@@ -0,0 +1,605 @@
...
+
+On teardown core rebuilds the edges from the references, prunes those naming
+resources that have left observed state, and asks the same `Decide` the live
+path uses. No function runs.
+
+The graph becomes readable by any tool with a `GET` on the XR, so `crossplane
+resource trace` and dashboards can render it without replaying a pipeline. And
+unlike references, `dependsOn` comes from function output, so an XR deleted
+before the pipeline next succeeds tears down by the previous graph which
+describes what was actually built.
```
[the remark this answers is not in the document]
Comment (stevendborrelli) [reply design/design-doc-composed-resource-ordering.md:469]: An XR whose references predate resourceName/dependsOn yields no edges, so it tears down exactly as it does today, in no particular order. The graph appears on the next reconcile that runs the pipeline, and teardown is ordered from then on. Enabling the feature changes nothing about an existing XR until then, and never leaves one stuck.

File: design/design-doc-composed-resource-ordering.md [hunk 3]
```diff
@@ -0,0 +1,605 @@
...
+Crossplane-runtime schemas to be updated:
+
+```yaml
+resourceRefs:
+- apiVersion: ec2.aws.upbound.io/v1beta1
+  kind: Subnet
+  name: my-xr-subnet-a-p4m9x
+  resourceName: subnet-a       # composition resource name
+  dependsOn: [vpc]             # composition resource names
+```
```
[the remark this answers is not in the document]
Comment (stevendborrelli) [reply design/design-doc-composed-resource-ordering.md:440]: Persisting lifecycle metadata with each edge — I don't think this one holds, and doing it would introduce a hang. `dependsOn` is read back in exactly one place: teardown, which builds its state with nothing desired, because the whole composition is going away. Create-before-destroy means "a replacement must exist and be ready before its predecessor goes away". Applied in that state it can never be satisfied — there is no replacement and never will be — so every such edge would block until the deadline. Dropping the lifecycle is what lets teardown terminate. That was an implicit choice before, and a poor one to leave implicit, so the doc now states it and says why.
